### PR TITLE
EntityMoveEvent

### PR DESCRIPTION
**Pull request by BillyGalbreath**
Adds the dreaded EntityMoveEvent everyone's been asking for since 2010.

Dont worry, the object creation is behind a handler size check. It will only create the event object if there is at least 1 listener to the event ;)

### CODE REVIEW

File: Spigot-Server-Patches/0582-EntityMoveEvent.patch [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
+index 76185f042dd773f267dc7574116cec7bd8a28ca8..399e4a67ca2433b4120dbfb5739730d1486ebb94 100644
+--- a/src/main/java/net/minecraft/server/EntityLiving.java
++++ b/src/main/java/net/minecraft/server/EntityLiving.java
+@@ -2801,6 +2801,20 @@ public abstract class EntityLiving extends Entity {
+ 
+         this.collideNearby();
+         this.world.getMethodProfiler().exit();
++        // Paper start
++        if (com.destroystokyo.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0) {
```
Comment (aikar) [inline Spigot-Server-Patches/0582-EntityMoveEvent.patch]: this should be cached so JIT can optimize this out. These method calls aren't simple getters.

**Review by aikar — CHANGES_REQUESTED** [review]

File: Spigot-API-Patches/0228-EntityMoveEvent.patch [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+Subject: [PATCH] EntityMoveEvent
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/event/entity/EntityMoveEvent.java b/src/main/java/com/destroystokyo/paper/event/entity/EntityMoveEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..87fb271141a9604d879455f029d5cc009637afe3
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/event/entity/EntityMoveEvent.java
+@@ -0,0 +1,95 @@
++package com.destroystokyo.paper.event.entity;
```
Comment (Proximyst) [inline Spigot-API-Patches/0228-EntityMoveEvent.patch]:
```suggestion
+package io.papermc.paper.event.entity;
```

**Review by Proximyst — DISMISSED** [review]

File: Spigot-Server-Patches/0582-EntityMoveEvent.patch [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
++        // Paper start
++        if (((WorldServer) world).hasEntityMoveEvent) {
++            if (lastX != locX() || lastY != locY() || lastZ != locZ() || lastYaw != yaw || lastPitch != pitch) {
++                Location from = new Location(world.getWorld(), lastX, lastY, lastZ, lastYaw, lastPitch);
++                Location to = new Location (world.getWorld(), locX(), locY(), locZ(), yaw, pitch);
++                com.destroystokyo.paper.event.entity.EntityMoveEvent event = new com.destroystokyo.paper.event.entity.EntityMoveEvent(getBukkitLivingEntity(), from, to.clone());
++                if (!event.callEvent()) {
++                    setLocation(from.getX(), from.getY(), from.getZ(), from.getYaw(), from.getPitch());
++                } else if (!to.equals(event.getTo())) {
++                    setLocation(to.getX(), to.getY(), to.getZ(), to.getYaw(), to.getPitch());
```
Comment (MrIvanPlays) [inline Spigot-Server-Patches/0582-EntityMoveEvent.patch]: shouldn't you use event.getTo() here?

File: Spigot-API-Patches/0228-EntityMoveEvent.patch (hunk 2 above)
Comment (BillyGalbreath) [reply Spigot-API-Patches/0228-EntityMoveEvent.patch]: I thought the plan was to move everything to the new package all at once so we don't split it up and create confusion

Comment (Proximyst) [reply Spigot-API-Patches/0228-EntityMoveEvent.patch]: That'll only be more work later on. We've already got some on the new package, and they're not moving in that case; little to lose by just using the new one at this point.

**Comment by BillyGalbreath** [comment]
Rebased

**Review by MiniDigger — DISMISSED** [review]

**Review by electronicboy — DISMISSED** [review]

**Comment by electronicboy** [comment]
provisionally approved, but, you broke it

**Comment by BillyGalbreath** [comment]
> provisionally approved, but, you broke it

Fixed. I missed a tokyo->io change :3

**Comment by BillyGalbreath** [comment]
Rebased (yet again)

**Review by electronicboy — APPROVED** [review]
My label is gone now, so, guess I got no other option :/

**Pull request merged**